### PR TITLE
Patch/#30 - various fixes for create group route

### LINKED ISSUE #3: Login Route

### LINKED ISSUE #30: Create Group Bug Fix
Creating a group with invalid/nonexistent usernames currently creates an empty group (users[] array is empty).

### PR DESCRIPTION
**Pull request by brandonLi8**
# Contributors

@brandonLi8 

# Relevant issue

Closes #30 (issue)

# Summary of change

1. Use https://pkg.go.dev/github.com/go-playground/validator for unique check (and added missing required, min length check)
2. Make sure that every username provided was found in the database, and return 400 if not
3. Make sure that the user who created the group (bearer) is inside the usernames.
4. Added unit test for create group route

# Testing/Verification

### Manual Tests
Login to get username1's token,
```
curl --location --request POST 'localhost:8080/auth/login' \
-H 'Content-Type: application/json' \
--data-raw '{
    "username": "username1",
    "password": "TestPassword"
}'
```
To test change 1:

```
curl --location --request POST 'localhost:8080/group' \
-H 'Content-Type: application/json' \
-H 'Authorization: Bearer <token>' \
--data-raw '{
    "usernames": ["username1", "username2",  "username2"]
}'
```

To test change 2:

```
curl --location --request POST 'localhost:8080/group' \
-H 'Content-Type: application/json' \
-H 'Authorization: Bearer <token>' \
--data-raw '{
    "usernames": ["username1", "username2",  "usernameThatDoesNotExist"]
}'
```

To test change 3:

```
curl --location --request POST 'localhost:8080/group' \
-H 'Content-Type: application/json' \
-H 'Authorization: Bearer <token for username1>' \
--data-raw '{
    "usernames": ["username2",  "username3"]
}'
```
### Unit Tests
- Added general unit test for create group
- Added spec for each of the manual tests above

### CODE REVIEW

File: controllers/group_controller.go [hunk 1]
```diff
@@ -67,12 +60,31 @@ func (ctrl *groupController) Create(context *gin.Context) {
...
 	//----------------------------------------------------------------------------------------
 
 	// Step 2: Find the id of each of the User models found.
-	var userIDs []primitive.ObjectID = make([]primitive.ObjectID, len(users))
+	// While iterating, make sure that the user who created the group is inside the userIDs.
+	userIDs := make([]primitive.ObjectID, len(users))
+	bearerUserFound := false
+
 	for i := 0; i < len(userIDs); i++ {
 		userIDs[i] = users[i].DefaultModel.IDField.ID
```
Comment (anjanbharadwaj) [inline controllers/group_controller.go]: Can change this to just `users[i].ID` instead?

File: controllers/group_controller.go [hunk 2]
```diff
@@ -67,12 +60,31 @@ func (ctrl *groupController) Create(context *gin.Context) {
...
+
+		if !bearerUserFound && userIDs[i].Hex() == context.MustGet("userID") {
+			bearerUserFound = true
+		}
+	}
+
+	// Make sure that the user who created the group is inside the usernames.
+	if !bearerUserFound {
+		context.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": "usernames must contain the bearer user"})
```
Comment (anjanbharadwaj) [inline controllers/group_controller.go]: Just a personal preference but would be nice to have standardized error messages (at least in terms of capitalizing first letter, punctuation at the end).

File: controllers/group_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	})
+}
+
+/*
+Gets the authorization (simulates logging in) for a specific user, based on the username.
+*/
+func LoginAuthUsername(context *gin.Context, username string) {
+	user := &models.User{}
+	mgm.Coll(user).First(bson.M{"username": username}, user)
+	context.Set("userID", user.DefaultModel.IDField.ID.Hex())
```
Comment (anjanbharadwaj) [inline controllers/group_controller_test.go]: same as before, users[i].ID?

**Review by anjanbharadwaj — CHANGES_REQUESTED** [review]
Mostly LGTM, just some small things here and there and you're good to go.

**Review by anjanbharadwaj — APPROVED** [review]

**Review by teekenzie — APPROVED** [review]

**Pull request merged**